sigal/plugins/encrypt/encrypt.py

```diff
@@ -0,0 +1,271 @@
+    logger.debug("encryption config: %s", config)
+
+    logger.info("starting encryption")
+    copy_assets(settings)
+    encrypt_files(settings, config, cache, albums, gallery.progressbar_target)
+    save_cache(settings, cache)
+
+def encrypt_files(settings, config, cache, albums, progressbar_target):
+    if settings["keep_orig"] and settings["orig_link"]:
+        logger.warning("Original images are symlinked! Encryption is aborted. Please set \"orig_link\" to False and restart gallery build.")
+        raise Abort
+
+    key = kdf_gen_key(config["password"], config["kdf_salt"], config["kdf_iters"])
+    gcm_tag = config["gcm_tag"].encode("utf-8")
+
+    medias = list(chain.from_iterable(albums.values()))
+    with progressbar(medias, label="%16s" % "Encrypting files", file=progressbar_target, show_eta=True) as medias:
+        for media in medias:
+            if media.type != "image":
+                logger.info("Skipping non-image file %s", media.filename)
+                continue
+
+            save_property(cache, media)
+            to_encrypt = get_encrypt_list(settings, media)
+
+            cacheEntry = cache[cache_key(media)]["encrypted"]
+            for f in to_encrypt:
+                if f in cacheEntry:
+                    logger.info("Skipping %s as it is already encrypted", f)
+                    continue
+
+                full_path = os.path.join(settings["destination"], f)
+                if encrypt_file(f, full_path, key, gcm_tag):
+                    cacheEntry.add(f)
+                else:
+                    # save the progress and abort the build if any image
+                    # fails to be encrypted
+                    save_cache(settings, cache)
+                    raise Abort
+
+    key_check_path = os.path.join(settings["destination"], 'static', 'keycheck.txt')
+    encrypt_file("keycheck.txt", key_check_path, key, gcm_tag)
+
+def encrypt_file(filename, full_path, key, gcm_tag):
+    with BytesIO() as outBuffer:
+        try:
+            with open(full_path, "rb") as infile:
+                encrypt(key, infile, outBuffer, gcm_tag)
+        except Exception as e:
+            logger.error("Encryption failed for %s: %s", filename, e)
+            return False
+        else:
+            logger.info("Encrypting %s...", filename)
+            try:
+                with open(full_path, "wb") as outfile:
+                    outfile.write(outBuffer.getbuffer())
+            except Exception as e:
+                logger.error("Could not write to file %s: %s", filename, e)
+                return False
+    return True
+
+def copy_assets(settings):
+    theme_path = os.path.join(settings["destination"], 'static')
+    copy(os.path.join(ASSETS_PATH, "decrypt.js"), theme_path, symlink=False, rellink=False)
+    copy(os.path.join(ASSETS_PATH, "keycheck.txt"), theme_path, symlink=False, rellink=False)
+    copy(os.path.join(ASSETS_PATH, "sw.js"), settings["destination"], symlink=False, rellink=False)
```

Comment:
Why `sw.js` is not in the static directory with the other static files ?

Reply:
As per [Service Worker specification](https://www.w3.org/TR/service-workers/#path-restriction), the worker script has no access to resources outside its directory tree. So putting `sw.js` in `static` will break it.

Reply:
Ok, thanks for the info!